libethash-cuda/CUDAMiner.cpp

```diff
@@ -478,24 +472,47 @@ void CUDAMiner::search(
             }
 
             // See if we got solutions in this batch
-            uint32_t found_count = std::min((unsigned)buffer->count, SEARCH_RESULTS);
+            uint32_t found_count = buffer->count;
             if (found_count)
             {
-
+                // save the solution(s), We'll submit them after all
+                buffer_cpy = &m_search_buf_cpy[current_index];
+                // the streams have been restarted.
+                if (found_count > SEARCH_RESULTS)
+                    found_count = SEARCH_RESULTS;
+                for (unsigned i = 0; i < found_count; i++)
+                    buffer_cpy->result[i] = ((search_results*)buffer)->result[i];
+                buffer_cpy->count = found_count;
                 buffer->count = 0;
-                uint64_t nonce_base = current_nonce - streams_batch_size;
+            }
+            // restart the stream on the next batch of nonces
+            if (!done)
+                run_ethash_search(
+                    s_gridSize, s_blockSize, stream, buffer, current_nonce, m_parallelHash);
+        }
+        // All streams are running. time for some bookeeping
+        updateHashRate(batch_size, s_numStreams);
+        for (current_index = 0; current_index < s_numStreams; current_index++)
+        {
+            buffer_cpy = &m_search_buf_cpy[current_index];
+            uint32_t found_count = buffer_cpy->count;
+            buffer_cpy->count = 0;
+            if (found_count)
+            {
+                uint64_t nonce_base =
+                    current_nonce - (2 * streams_batch_size) + (current_index * batch_size);
```

Comment:
This really goes beyond my understanding. That "2" shouldn't be the number of streams ?
Why constant ?

Reply:
Because stream_batch_size = s_numStreams * batch_size, and the result is from 4 batches ago.

Reply:
And why the double of `streams_batch_size` which is already `batch_size * s_numStreams` ?
I really don't get it ... apparently you're altering the minernonce

Reply:
> and the result is from 4 batches ago.

I dont think it's correct. Reading your code the result you're processing is from last completed batch where _n_ streams ran concurred to it.